docs/develop/dependency-injection.md

```diff
@@ -0,0 +1,144 @@
+	// The implementation is elided here, and generated at compile-time by Encore.
+	return nil
+}
+```
+
+These functions are generated in order to allow other services to keep calling your
+APIs as package-level functions, in the same way as before: `email.Send(...)`.
+Other services do not have to care about whether you're using Dependency Injection
+internally. You must always use these generated package-level functions for making API calls.
+
+<Callout type="important">
+
+Encore will automatically generate these files and keep them up to date
+whenever your code changes. There is no need to manually invoke anything
+to regenerate this code.
+
+</Callout>
+
+Encore adds all `encore.gen.go` files to your `.gitignore` since you typically
+don't want to commit them to your repository; doing so ends up creating
+a lot of unnecessary merge conflicts.
+
+However, in some cases when running third-party linters in a CI/CD environment
+it can be helpful to generate these wrappers to make the linter happy.
+You can do that by invoking `encore gen wrappers`.
+
+### Mocking dependencies
+
+If you wish to mock out the SendGrid client for testing purposes you can change the
+field to an interface:
+
+```go
+type sendgridClient interface {
+	SendEmail(...) // a hypothetical signature, for illustration purposes
+}
+
+//encore:service
+type Service struct {
+    sendgridClient sendgridClient
+}
+```
+
+Then during your tests you can instantiate the service object by hand:
+```go
+func TestFoo(t *testing.T) {
+    svc := &Service{sendgridClient: &myMockClient{}}
+    // ...
+}
+```
+
+### Graceful Shutdown
+
+When defining a service struct, Encore supports notifying
+your service when it's time to gracefully shut down. This works
+by having your service struct implement the method
+`func (s *Service) Shutdown(force context.Context)`.
+
+
+If that method exists, Encore will call it when it's time to begin
+gracefully shutting down. Initially the shutdown is in "graceful mode",
+which means that you have a few seconds to complete ongoing work.
+
+<Callout type="important">
+
+Encore automatically handles graceful shutdown of all Encore-managed
+functionality, such as HTTP servers, database connection pools,
+Pub/Sub message receivers, distributed tracing recorders, and so on.
+The graceful shutdown functionality is provided if you have additional,
+non-Encore-related resources that need graceful shutdown.
+
+</Callout>
+
+The provided `force` context is canceled when the graceful shutdown window
+is over, and it's time to forcefully shut down. The amount of time you have
+from when `Shutdown` is called to when forceful shutdown begins depends on the
+cloud provider and the underlying infrastructure, but typically is anywhere
+from 5-30 seconds.
+
+Note that graceful shutdown in Encore is *cooperative*: Encore will wait indefinitely
+for your `Shutdown` method to return. If your `Shutdown` method does not return promptly
```

Comment:
So the way I read this is "don't return from your `Shutdown` method until your system has completed graceful shutdown".

Which I don't think is the case, I think it's "You can immediately return after triggering a graceful shutdown".

However if we expect customers to wait for the graceful shutdown to finish we should be clear about that. Actually I think maybe we should be clear either way here.

Reply:
It does wait until you return, but I thought that was pretty clear from the docs. Not sure how to make it clearer